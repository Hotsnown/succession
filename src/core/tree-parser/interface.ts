export interface Tree {
    [key: string]: Node
}

export interface Node {
    id: string;
    name: string;
    partners?: string[];
    children?: HouseHold;
}

export interface HouseHold {
    [key: string]: string[]
}

export interface Data {
}

export interface Family {
    member_id: string;
    childs: string[];
    data: Data;
}