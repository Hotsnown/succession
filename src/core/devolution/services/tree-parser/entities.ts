/* prettier-ignore */
/*eslint-disable*/

export class Output {
    de_cujus: string
    family: RawMember[]

    constructor(de_cujus: string){
        this.de_cujus = de_cujus
        this.family = []
    }

    appendFamily(member_id: string, childs: string[], status: Status, spouse?: string[]) {
        this.family.push({
            member_id: member_id, 
            childs: childs,
            attributes: {
                status: status,
                spouse: spouse || 'without spouse'
            }})
    }
}

interface RawMember {
    member_id: string;
    childs: string[];
    attributes: Attributes;
    index?: number
}

type Status = 'valid' | 'invalid'

interface Attributes {
    status: Status
    spouse: string[] | 'without spouse'
}