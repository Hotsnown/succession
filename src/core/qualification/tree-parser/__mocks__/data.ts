const simpsonsTree = [
  {
    "id": "unknown",
    "name": "???",
    "partners": [],
    "children": {}
  },
  {
    "id": "mona",
    "name": "Mona Penelope Simpson (née Olsen)",
    "partners": [],
    "children": {}
  },
  {
    "id": "herb",
    "name": "Herbert (Herb) Powell",
    "partners": [],
    "children": {}
  },
  {
    "id": "marge",
    "name": "Marjorie (Marge) Simpson (née Bouvier)",
    "partners": [],
    "children": {}
  },
  {
    "id": "bart",
    "name": "Bartholomew (Bart) JoJo Simpson",
    "partners": [],
    "children": {}
  },
  {
    "id": "millhouse",
    "name": "Millhouse Van Houten",
    "partners": [],
    "children": {}
  },
  {
    "id": "millhouse_jr",
    "name": "Millhouse Van Houten Jr.",
    "partners": [],
    "children": {}
  },
  {
    "id": "lisa",
    "name": "Lisa Marie Simpson",
    "partners": [
      {
        "id": "millhouse",
        "name": "Millhouse Van Houten",
        "partners": [],
        "children": {}
      }
    ],
    "children": {
      "millhouse": [
        {
          "id": "millhouse_jr",
          "name": "Millhouse Van Houten Jr.",
          "partners": [],
          "children": {}
        }
      ]
    }
  },
  {
    "id": "maggie",
    "name": "Margaret (Maggie) Eve Simpson",
    "partners": [],
    "children": {}
  },
  {
    "id": "homer",
    "name": "Homer Jay Simpson",
    "partners": [
      {
        "id": "marge",
        "name": "Marjorie (Marge) Simpson (née Bouvier)",
        "partners": [],
        "children": {}
      }
    ],
    "children": {
      "marge": [
        {
          "id": "bart",
          "name": "Bartholomew (Bart) JoJo Simpson",
          "partners": [],
          "children": {}
        },
        {
          "id": "lisa",
          "name": "Lisa Marie Simpson",
          "partners": [
            {
              "id": "millhouse",
              "name": "Millhouse Van Houten",
              "partners": [],
              "children": {}
            }
          ],
          "children": {
            "millhouse": [
              {
                "id": "millhouse_jr",
                "name": "Millhouse Van Houten Jr.",
                "partners": [],
                "children": {}
              }
            ]
          }
        },
        {
          "id": "maggie",
          "name": "Margaret (Maggie) Eve Simpson",
          "partners": [],
          "children": {}
        }
      ]
    }
  },
  {
    "id": "abe",
    "name": "Abraham J. (Grandpa) Simpson",
    "partners": [
      {
        "id": "unknown",
        "name": "???",
        "partners": [],
        "children": {}
      },
      {
        "id": "mona",
        "name": "Mona Penelope Simpson (née Olsen)",
        "partners": [],
        "children": {}
      }
    ],
    "children": {
      "unknown": [
        {
          "id": "herb",
          "name": "Herbert (Herb) Powell",
          "partners": [],
          "children": {}
        }
      ],
      "mona": [
        {
          "id": "homer",
          "name": "Homer Jay Simpson",
          "partners": [
            {
              "id": "marge",
              "name": "Marjorie (Marge) Simpson (née Bouvier)",
              "partners": [],
              "children": {}
            }
          ],
          "children": {
            "marge": [
              {
                "id": "bart",
                "name": "Bartholomew (Bart) JoJo Simpson",
                "partners": [],
                "children": {}
              },
              {
                "id": "lisa",
                "name": "Lisa Marie Simpson",
                "partners": [
                  {
                    "id": "millhouse",
                    "name": "Millhouse Van Houten",
                    "partners": [],
                    "children": {}
                  }
                ],
                "children": {
                  "millhouse": [
                    {
                      "id": "millhouse_jr",
                      "name": "Millhouse Van Houten Jr.",
                      "partners": [],
                      "children": {}
                    }
                  ]
                }
              },
              {
                "id": "maggie",
                "name": "Margaret (Maggie) Eve Simpson",
                "partners": [],
                "children": {}
              }
            ]
          }
        }
      ]
    }
  }
]

export default simpsonsTree