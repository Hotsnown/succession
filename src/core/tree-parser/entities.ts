export class Output {
    de_cujus: string
    family: Family[]

    constructor(de_cujus: string){
        this.de_cujus = de_cujus
        this.family = []
    }

    jsonify() {
        return JSON.stringify(
            {
                de_cujus: this.de_cujus, 
                family: this.family
            })
    }

    appendFamily(member_id: string, childs: string[]) {
        this.family.push({member_id: member_id, childs: childs, data: {}})
    }
}

interface Family {
    member_id: string;
    childs: string[];
    data: Data;
}

interface Data {
}