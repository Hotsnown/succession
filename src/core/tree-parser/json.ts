import { Node, RootObject } from "./interface";

export class Output {
    de_cujus: string
    family: Family[]

    constructor(de_cujus: string, family: Family[]){
        this.de_cujus = de_cujus
        this.family = family
    }

    jsonify() {
        return JSON.stringify(
            {
                de_cujus: this.de_cujus, 
                family: this.family
            })
    }

    appendFamily(member_id: string, childs: string[]) {
        this.family.push(new Family(member_id, childs))
    }
}

class Family {
    member_id: string;
    childs: string[];
    data: Data;
    
    constructor (member_id: string, childs: string[]) {
        this.member_id = member_id
        this.childs = childs
        this.data = {}
    }
}

interface Data {
}