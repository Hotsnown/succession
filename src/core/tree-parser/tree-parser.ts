interface Tree {
    [key: string]: Node
}

interface Node {
    id: string;
    name: string;
    partners?: string[];
    children?: HouseHold;
}

interface HouseHold {
    [key: string]: string[]
}

export function treeParser(tree: Tree) {
    let facts: string[] = []

    Object.entries(tree)
        .map(
            node => {

                let parent = node[0]
                let houseHold = node[1]["children"]

                if (houseHold) {
                    if (hasMultipleChildren(houseHold)) {
                        getChildren(houseHold)
                            .map(childrens => extractMultipleChildren(childrens, facts, parent))
                    }
                    else {
                        getChildren(houseHold)
                            .map(extractSingleChildren(facts, parent))
                    }
                }
            })

    return facts
}

function getChildren(houseHold: HouseHold) {
    return Object.values(houseHold)
}

function hasMultipleChildren(houseHold: HouseHold) {
    return getChildren(houseHold).length
}

function extractSingleChildren(facts: string[], parent: string): (value: string[], index: number, array: string[][]) => number {
    return children => facts.push(`parent(${parent}, ${children}).`)
}

function extractMultipleChildren(childrens: string[], facts: string[], parent: string): number[] {
    return childrens
        .map((children: any) => facts.push(`parent(${parent}, ${children}).`))
}
