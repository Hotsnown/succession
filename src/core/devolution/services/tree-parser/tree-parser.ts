/* prettier-ignore */
/*eslint-disable*/

import { HouseHold } from './interface'
import { Output } from './entities'

interface Tree {
    id: string;
    name: string;
    status: Status
    partners: Tree;
    children: Tree
}

type Status = 'valid' | 'invalid'

//TODO : change to traverse tree

function treeParser(tree: Tree, deCujus: string): Output {
    let facts: Output = new Output(deCujus)

    Object.entries(tree)
        .map(parseNode(facts))

    return facts
}

export default treeParser

function parseNode(facts: Output) {
    return (node: any) => {
        let houseHold = node[1].children
        if (houseHold !== undefined && !isEmpty(houseHold) ) {
            if (hasMultipleChildren(houseHold)) {
                extractMultipleChildren(getChildrenString(houseHold), facts, node[1].id, node[1].status)
            }
            else {
                getChildrenArray(houseHold)
                    .map(extractSingleChildren(facts, node[1].id, node[1].status))
            }
        }
        else {
            facts.appendFamily(node[1].id, [], node[1].status)
        }
    }
}

function getChildrenString(houseHold: HouseHold): string[] {
    return Object.values(houseHold).flat(1).map((c: any) => c.id)
}

function getChildrenArray(houseHold: HouseHold) {
    return Object.values(houseHold)
}

function hasMultipleChildren(houseHold: HouseHold) {
    return getChildrenArray(houseHold).length
}

function extractMultipleChildren(childrens: string[], facts: Output, parent: string, status: Status) {
    facts.appendFamily(parent, childrens, status)
}

function extractSingleChildren(facts: Output, parent: string, status: Status) {
    return (children: any) =>  facts.appendFamily(parent, children.id, status)
}

function isEmpty(obj: Object) {
    for(var key in obj) {
        if(obj.hasOwnProperty(key))
            return false;
    }
    return true;
}