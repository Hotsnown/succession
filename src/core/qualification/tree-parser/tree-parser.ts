import { Tree, Node, HouseHold } from './interface'
import { Output } from './entities'

function treeParser(tree: any, deCujus: string) {
    let facts: Output = new Output(deCujus)

    Object.entries(tree)
        .map(parseNode(facts))

    return facts
}

export default treeParser

function parseNode(facts: Output) {
    return (node: any) => {
        let parent = node[0]
        let houseHold = node[1].children
        if (houseHold !== undefined && !isEmpty(houseHold) ) {
            if (hasMultipleChildren(houseHold)) {
                extractMultipleChildren(getChildrenString(houseHold), facts, parent)
            }
            else {
                getChildrenArray(houseHold)
                    .map(extractSingleChildren(facts, parent))
            }
        }
        else {
            facts.appendFamily(parent, [])
        }
    }
}

function getChildrenString(houseHold: HouseHold): string[] {
    return Object.values(houseHold).flat(1).map((c: any) => c.id)
}

function getChildrenArray(houseHold: HouseHold) {
    return Object.values(houseHold)
}

function hasMultipleChildren(houseHold: HouseHold) {
    return getChildrenArray(houseHold).length
}

function extractMultipleChildren(childrens: string[], facts: Output, parent: string) {
    facts.appendFamily(parent, childrens)
}

function extractSingleChildren(facts: Output, parent: string) {
    return (children: any) =>  facts.appendFamily(parent, children.id)
}

function isEmpty(obj: Object) {
    for(var key in obj) {
        if(obj.hasOwnProperty(key))
            return false;
    }
    return true;
}