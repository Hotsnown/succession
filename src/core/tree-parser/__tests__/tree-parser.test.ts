import { treeParser } from '../tree-parser'
import { simpsonsTree } from './data'

it('', () => {
    expect(treeParser(simpsonsTree)).toMatchSnapshot()
})