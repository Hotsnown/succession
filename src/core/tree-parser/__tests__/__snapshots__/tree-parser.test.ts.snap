// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  "parent(abe, herb).",
  "parent(abe, homer).",
  "parent(homer, bart).",
  "parent(homer, lisa).",
  "parent(homer, maggie).",
  "parent(lisa, millhouse_jr).",
]
`;
